app/src/main/kotlin/com/wire/android/mapper/MessageContentMapper.kt

```diff
@@ -121,24 +135,28 @@ class MessageContentMapper @Inject constructor(
         }
     ).let { messageBody -> UIMessageContent.TextMessage(messageBody = messageBody) }
 
-    suspend fun toAsset(
-        conversationId: QualifiedID,
-        messageId: String,
-        assetContent: AssetContent
-    ) = with(assetContent) {
-        val (imgWidth, imgHeight) = when (val md = metadata) {
-            is AssetContent.AssetMetadata.Image -> md.width to md.height
-            else -> 0 to 0
-        }
-        if (remoteData.assetId.isNotEmpty()) {
+    suspend fun toUIMessageContent(assetMessageData: AssetMessageData, message: Message, scope: CoroutineScope) =
+        with(assetMessageData.assetMessageContent) {
             when {
                 // If it's an image, we download it right away
-                isValidImage(mimeType) && imgWidth.isGreaterThan(0) && imgHeight.isGreaterThan(0) -> UIMessageContent.ImageMessage(
-                    assetId = AssetId(remoteData.assetId, remoteData.assetDomain.orEmpty()),
-                    imgData = imageRawData(conversationId, messageId),
-                    width = imgWidth,
-                    height = imgHeight
-                )
+                assetMessageData.isValidImage() -> {
+                    var imageRawData: ByteArray? = null
+                    val job = scope.launch(dispatcherProvider.io()) {
+                        val imageData = imageRawData(message.conversationId, message.id)
+                        withContext(Dispatchers.Main) { imageRawData = imageData }
```

Comment:
`withContext` can return a value so we can simplify it and get rid of nullable var 😉

Reply:
Also, shouldn't we use `dispatcherProvider.main()`? 🤔